src/main.rs

```diff
@@ -71,203 +84,227 @@ impl Player {
-}
-
-impl Room {
-    fn new(id: usize) -> Self {
-        let default_room = Room::default();
-
-        Room {
-            id,
-            ..default_room
-        }
-    }
-
-    fn missing_neighbours(&self) -> usize {
-        self.neighbours.iter().filter(|n| n.get().is_none()).count()
-    }
-
-    fn neighbour_ids(&self) -> Vec<usize> {
-        self.neighbours.iter()
-            .filter(|n| n.get().is_some())
-            .map(|n| n.get().unwrap())
-            .collect()
+impl Default for RoomContents {
+    fn default() -> Self {
+        Empty
     }
 }
 
-//////////
-// MAZE //
-//////////
+/// Room description.
+#[derive(Default, Clone, Copy)]
+struct Room {
+    /// The indices of neighboring rooms.
+    neighbors: [usize; ROOM_NEIGHBORS],
+    /// Possible danger in the room.
+    contents: RoomContents,
+    /// Wumpus in the room.
+    wumpus: bool,
+}
 
-#[derive(Debug)]
+/// The Maze, including RNG state.
 struct Maze {
-    rooms: Vec<Room>,
-    rng: Rc<RefCell<ThreadRng>>,
+    /// Room list.
+    rooms: [Room; MAZE_ROOMS],
+    /// RNG state.
+    rng: ThreadRng,
 }
 
 impl Maze {
-    // Builds a vector of interconnected rooms by looping each room and finding three other rooms
-    // to link to it
-    fn new(rng: Rc<RefCell<ThreadRng>>) -> Self {
-        let rooms: Vec<Room> = (0..MAZE_ROOMS)
-            .map(|idx| Room::new(idx as usize))
-            .collect();
-
-        // for each room, see how many missing neighbours there are and them
-        for room in rooms.iter() {
-            for idx in 0..room.neighbours.len() {
-                if room.neighbours[idx].get().is_none() {
-                    // a suitable neighbour room is a room that is not the current one, not yet
-                    // linked to the current one and that still has available neighbours
-                    let neighbour = rooms.iter()
-                        .find(|n| true &&
-                            n.id != room.id
-                            && n.missing_neighbours() != 0
-                            && !room.neighbour_ids().contains(&n.id)
-                        )
-                        .expect("Cannot find a suitable neighbour");
-
-                    room.neighbours[idx].set(Some(neighbour.id));
-
-                    // link back the current room to the neighbour so that Room #0 has a link to
-                    // Room #1 and Room #1 has a link to Room #0
-                    for i in 0..ROOM_NEIGHBOURS {
-                        if neighbour.neighbours[i].get().is_none() {
-                            neighbour.neighbours[i].set(Some(room.id));
-                            break;
-                        }
-                    }
-                }
-            }
-        }
-
-        let mut maze = Maze {
-            rooms,
-            rng,
+    // List of adjacencies used to wire up the dodecahedron.
+    // https://stackoverflow.com/a/44096541/364875
+    const ADJS: [[usize; 3]; 20] = [
```

Comment:
I really don't want the tunnels between rooms to be hard-coded.
Some solutions on Rosetta Code do it, some other have an algorithm to compute the connections at run time. I deliberately wanted to avoid having the connects hard-coded since

1. from an algorithmic point of view, computing the maze is the most interesting part of the code
2. it makes it dead easy to implement "difficulty levels" by tweaking constants: more pits, more bats, more rooms, more connections between the rooms (want en icosahedron? just set the number of rooms to 12 and the number of connections between the rooms to 5).

The proposed `Maze::new` code aims to be a simplification, but in the end the code is harder or similarly complex and above all so much less flexible.

Reply:
If you want to go back to maze generation, you need to ensure that the maze is fully-connected. Did your old code guarantee that? It looked to me like it could end up with a situation where the Wumpus was unreachable because it was in a different connected component of the Maze than the player?

(In any case, if you're not going to have an icosahedron you don't have an implementation of Hunt the Wumpus in my opinion. It is one of the best-known features of the game.)

Reply:
I'm afraid I don't agree that the new code is as complex. The old code involved `Option` types in a fundamental way, and meant that the maze representation wasn't uniform. A whole bunch of the surrounding code was simplified by the invariant that every room has exactly three neighbors with no `Option` type to worry about. If you want to break that invariant, I'd suggest neighbors be a `Vec<usize>` instead of an `[Option<usize>; 3]`. This would preserve most of the simplification.